next-frontend/src/providers/PermissionProvider.tsx

```diff
@@ -0,0 +1,73 @@
+  canViewDelegateAdminPage: (competition: string) => boolean
+  canViewDelegateReport: (competition: string) => boolean
+  canCreateGroup: (group: string) => boolean
+  canEditGroup: (group: string) => boolean
+  canReadGroupCurrent: (group: string) => boolean
+  canReadGroupPast: (group: string) => boolean
+  canRequestToEditProfile: (profile: string) => boolean
+}
+
+const PermissionContext = createContext<PermissionContext | null>(null);
+
+export const usePermissions = () => useContext(PermissionContext);
+
+const allOrSpecificScope = (item: string, scope: components["schemas"]["CompetitionPermissions"]) => {
+  return scope === "*" || scope.includes(item)
+}
+
+export default function PermissionProvider({ children }: { children: React.ReactNode }){
+  const { data: session } = useSession();
+  const api = useAPI();
+  const { data: request, isLoading } = useQuery({
+    enabled: Boolean(session),
+    queryKey: ["permissions", session?.user?.id],
+    queryFn: () => api.GET("/users/me/permissions")
+  })
+
+  const permissions: PermissionContext = useMemo(() => {
+    const rawPermissions = request?.data
+
+    return {
+      permissions: rawPermissions,
+      canAccessPanel: panel => Boolean(rawPermissions && rawPermissions.can_access_panels.scope.includes(panel)),
+      canAdministerCompetition: competition => Boolean(rawPermissions && allOrSpecificScope(competition, rawPermissions.can_administer_competitions.scope)),
+      canAttendCompetition: competition => Boolean(rawPermissions && allOrSpecificScope(competition, rawPermissions.can_attend_competitions.scope)),
+      canOrganizeCompetitions: competition => Boolean(rawPermissions && allOrSpecificScope(competition, rawPermissions.can_organize_competitions.scope)),
+      canEditDelegateReport: competition => Boolean(rawPermissions && allOrSpecificScope(competition, rawPermissions.can_edit_delegate_report.scope)),
+      canViewDelegateReport: competition => Boolean(rawPermissions && allOrSpecificScope(competition, rawPermissions.can_view_delegate_report.scope)),
+      canViewDelegateAdminPage: competition => Boolean(rawPermissions && allOrSpecificScope(competition, rawPermissions.can_view_delegate_admin_page.scope)),
+      canCreateGroup: group => Boolean(rawPermissions && rawPermissions.can_create_groups.scope.includes(group)),
+      canEditGroup: group => Boolean(rawPermissions && rawPermissions.can_edit_groups.scope.includes(group)),
+      canReadGroupCurrent: group => Boolean(rawPermissions && rawPermissions.can_read_groups_current.scope.includes(group)),
+      canReadGroupPast: group => Boolean(rawPermissions && rawPermissions.can_read_groups_past.scope.includes(group)),
+      canRequestToEditProfile: profile => Boolean(rawPermissions && allOrSpecificScope(profile, rawPermissions.can_request_to_edit_others_profile.scope)),
+    }
+  }, [request])
+
+  if(isLoading) {
+    return (
+      <p>Loading...</p>
+    )
+  }
```

Comment:
Still not too wild about rendering a DOM node while loading. Shouldn't the `Boolean(rawPermissions && actualPermissionCheck(...))` trick be enough?

Reply:
it might create jumpy UIs were it renders some elements differently a second time if we don't

Reply:
How so?

Reply:
Besides potential UI jumping around, passing 'incorrect' permissions to components seems bad. There could also be (potentially, in the future) side-effects depending on these permissions, which would be triggered with the wrong values.